### DISCUSSION TITLE
Tasmota + DHT22 Sensor on Wemos D1 Mini ESP8266 - problem with negativ values

### DISCUSSION THREAD

**Discussion by ReelHiJacker** [arendst/Tasmota#20514] [Q&A] [answered]
I've a big problem with my DHT22 Sensors and Tasmota on a Wemos D1 Mini. Temperatures above 0 Degree are working perfect, but when the temperatur drops below 0 it shows values above 400° Celsius.
I tried different DHT22 Sensors from 3 different sellers (also the DHT22 Shield), different Tasmota Versions (9.5.0, 13.2.0, 13.3.03), different Wemos D1 Minis nothing changes. All settings had the problem with the Temperatur below 0, and all show values above 400 degree. 

The DHT22 is connected to D1 (GPIO 5) as AM2301 (and i also tried diffent GPIO's). 

I've found an old issue here in github (#14173): there where values around -3200 degree below 0 but this was closed and corrected.

Can you help me?

![Bildschirmfoto 2024-01-16 um 06 23 27](https://github.com/arendst/Tasmota/assets/52639588/9f9f381f-e722-4c80-8504-66e904eca523)
![Bildschirmfoto 2024-01-16 um 06 23 35](https://github.com/arendst/Tasmota/assets/52639588/2c920968-5515-40b0-bab9-056fe6a78d24)

**Comment by sfromis** [comment]
Just did a test with DHT22 and a freezing element, both on ESP32 and ESP8266. No signs of trouble. 
![image](https://github.com/arendst/Tasmota/assets/47082390/a13bf359-0a65-4cad-a6ec-94b39697ad00)

However the DHT11 is known not to handle freezing temperatures.

```
15:11:04.383 DHT: Pin13 cycles (80/80)  68 35 68 36 74 35 74 35 75 35 74 35 75 102 74 100 94 102 74 35 ..
15:11:04.385 DHT: Pin13 read 038B802432
15:11:08.386 DHT: Pin13 cycles (80/80)  68 35 68 36 74 35 74 35 75 35 74 35 75 102 74 100 94 102 74 35 ..
15:11:08.388 DHT: Pin13 read 038A802431
15:11:12.344 DHT: Pin13 cycles (80/80)  69 35 68 35 74 36 74 35 74 35 75 35 74 102 74 101 93 102 74 36 ..
15:11:12.346 DHT: Pin13 read 0389802531
15:11:16.354 DHT: Pin13 cycles (80/80)  68 35 68 36 74 35 74 35 75 35 74 35 75 102 74 100 94 102 74 35 ..
15:11:16.355 DHT: Pin13 read 0388802631
15:11:20.376 DHT: Pin13 cycles (80/80)  69 35 68 35 74 36 74 35 74 35 75 35 74 102 74 101 93 102 74 36 ..
15:11:20.378 DHT: Pin13 read 0388802631
15:11:24.357 DHT: Pin13 cycles (80/80)  69 35 68 35 74 36 74 35 74 35 75 35 74 102 74 101 93 102 74 36 ..
15:11:24.359 DHT: Pin13 read 0387802731
15:11:28.358 DHT: Pin13 cycles (80/80)  68 35 68 36 74 35 74 35 75 35 74 35 75 102 74 100 94 102 74 35 ..
15:11:28.360 DHT: Pin13 read 0385802830
15:11:32.360 DHT: Pin13 cycles (80/80)  68 35 68 36 74 35 74 35 75 35 74 35 75 102 74 100 94 102 74 35 ..
15:11:32.362 DHT: Pin13 read 0385802830
```

**Reply by ReelHiJacker** [reply]
These are the one i'm actual using
![IMG_5330](https://github.com/arendst/Tasmota/assets/52639588/f4d3e17c-3f94-4e3f-a378-700e4c281fc6)

My first tought was that this is a problem with the dht22's (i'm using other ones first, with the same problem). Then i bougth the ones with the shield and than this one (all from different sellers). I've got 3 of this above ones and all have the same behaviour also as the other one's i'm using first. So i don't believe it is my dht's?!

As i can see, my dth22 sending the 407 degree value in hex format as yours sending values above 30000... so for my unterstanding: the hex values i see, are these the values comming direct from the sensor or is tasmota generating the hex format?

Don't know what to do...

**Reply by sfromis** [reply]
Well, with positive values, I'm seeing the 2nd 16-bit number as simply degrees C with one decimal. With negative values, it looks like someone decided to not use two's complement representation, but just set the high-order bit for negative. Don't read this as values above 32768, as it more looks like 1 bit sign indicator followed by 15 bit data.

Your temperature values of around hex 0fe5 does not look like an even plausible representation of a negative number, but some other way of making a value. Looks like your "DHT22" is not DHT22-compatible then, unless it is somehow a newer and different revision.

Do you have a data sheet to support your values as being compliant? Since I have a working DHT22, I'm assuming this to be compliant. However, the trouble I've seen around one-wire sensors has made me put it in the drawer, as I've found I2C sensors very straightforward to deal with instead.

**Comment by ReelHiJacker** [comment, marked as answer]
So, i ordered some new DHT22 and now i was looking that these are real DHT22 with a print on it. And voila they are working right of the box! So the problem were the DHT22...
